src/hotspot/share/runtime/trimNativeHeap.cpp

```diff
@@ -0,0 +1,271 @@
+            ml.wait(wait_ms);
+          } else if (at_or_nearing_safepoint()) {
+            times_safepoint ++;
+            const int64_t wait_ms = MIN2<int64_t>(TrimNativeHeapInterval, safepoint_poll_ms);
+            ml.wait(wait_ms);
+          }
+
+          if (_stop) return;
+
+          tnow = now();
+        }
+      }
+
+      log_trace(trimnative)("Times %u suspended, %u timed, %u safepoint",
+                            times_suspended, times_waited, times_safepoint);
+
+      execute_trim_and_log(tnow);
+    }
+  }
+
+  // Execute the native trim, log results.
+  void execute_trim_and_log(double t1) {
+    assert(os::can_trim_native_heap(), "Unexpected");
+
+    os::size_change_t sc = { 0, 0 };
+    LogTarget(Info, trimnative) lt;
+    const bool logging_enabled = lt.is_enabled();
+
+    // We only collect size change information if we are logging; save the access to procfs otherwise.
+    if (os::trim_native_heap(logging_enabled ? &sc : nullptr)) {
+      _num_trims_performed++;
+      if (logging_enabled) {
+        double t2 = now();
+        if (sc.after != SIZE_MAX) {
+          const size_t delta = sc.after < sc.before ? (sc.before - sc.after) : (sc.after - sc.before);
+          const char sign = sc.after < sc.before ? '-' : '+';
+          log_info(trimnative)("Periodic Trim (" UINT64_FORMAT "): " PROPERFMT "->" PROPERFMT " (%c" PROPERFMT ") %.3fms",
+                               _num_trims_performed,
+                               PROPERFMTARGS(sc.before), PROPERFMTARGS(sc.after), sign, PROPERFMTARGS(delta),
+                               to_ms(t2 - t1));
+        } else {
+          log_info(trimnative)("Periodic Trim (" UINT64_FORMAT "): complete (no details) %.3fms",
+                               _num_trims_performed,
+                               to_ms(t2 - t1));
+        }
+      }
+    }
+  }
+
+public:
+
+  NativeHeapTrimmerThread() :
+    _lock(new (std::nothrow) PaddedMonitor(Mutex::nosafepoint, "NativeHeapTrimmer_lock")),
+    _stop(false),
+    _suspend_count(0),
+    _num_trims_performed(0)
+  {
+    set_name("Native Heap Trimmer");
+    if (os::create_thread(this, os::vm_thread)) {
+      os::start_thread(this);
+    }
+  }
+
+  void suspend(const char* reason) {
+    assert(NativeHeapTrimmer::enabled(), "Only call if enabled");
+    uint16_t n = 0;
+    {
+      MonitorLocker ml(_lock, Mutex::_no_safepoint_check_flag);
+      n = inc_suspend_count();
+      // No need to wakeup trimmer
+    }
+    log_debug(trimnative)("Trim suspended for %s (%u suspend requests)", reason, n);
+  }
+
+  void resume(const char* reason) {
+    assert(NativeHeapTrimmer::enabled(), "Only call if enabled");
+    uint16_t n = 0;
+    {
+      MonitorLocker ml(_lock, Mutex::_no_safepoint_check_flag);
+      n = dec_suspend_count();
+      if (n == 0) {
+        ml.notify_all(); // pause end
+      }
+    }
+    log_debug(trimnative)("Trim resumed after %s (%u suspend requests)", reason, n);
```

Comment:
Can you do it like I did in https://github.com/openjdk/jdk/files/12043977/trimnative-shipilev-2.patch ?

We don't say "resumed" if it really want not resumed due to non-zero suspend count.